src/tensors/dbcsr_tensor_api_c.F

```diff
@@ -0,0 +1,1231 @@
+      sizes_of_arrays, sum_of_arrays, array_sublist, get_arrays, get_ith_array, array_eq_i
+   USE dbcsr_tensor, ONLY: &
+      dbcsr_t_contract, dbcsr_t_get_block, dbcsr_t_get_stored_coordinates, dbcsr_t_put_block, &
+      dbcsr_t_reserve_blocks, dbcsr_t_copy_matrix_to_tensor, dbcsr_t_copy, &
+      dbcsr_t_copy_tensor_to_matrix
+   USE dbcsr_tensor_block, ONLY: &
+      dbcsr_t_iterator_blocks_left, dbcsr_t_iterator_next_block, dbcsr_t_iterator_start, &
+      dbcsr_t_iterator_stop, dbcsr_t_iterator_type
+   USE dbcsr_tensor_types, ONLY: &
+      dbcsr_t_create, dbcsr_t_destroy, dbcsr_t_distribution_destroy, dbcsr_t_distribution_new, &
+      dbcsr_t_distribution, dbcsr_t_distribution_type, dbcsr_t_nd_mp_comm, dbcsr_t_nd_mp_free, dbcsr_t_type, &
+      dbcsr_t_pgrid_type, dbcsr_t_pgrid_create, dbcsr_t_pgrid_destroy, dbcsr_t_set, dbcsr_t_filter, &
+      dbcsr_t_mp_environ_pgrid => mp_environ_pgrid, dbcsr_t_blk_sizes, dbcsr_t_get_info, &
+      dbcsr_t_finalize, dbcsr_t_scale, dbcsr_t_get_nze, dbcsr_t_get_nze_total, &
+      dbcsr_t_get_num_blocks, dbcsr_t_get_num_blocks_total, dbcsr_t_clear, dbcsr_t_get_data_type, &
+      mp_environ_pgrid, dbcsr_t_pgrid_create_expert
+   USE dbcsr_tensor_split, ONLY: &
+      dbcsr_t_split_blocks
+   USE dbcsr_kinds, ONLY: &
+      ${uselist(dtype_float_prec)}$, default_string_length, int_8
+   USE dbcsr_data_types, ONLY: dbcsr_scalar_type
+   USE dbcsr_data_methods, ONLY: dbcsr_scalar
+   USE dbcsr_machine, ONLY: default_output_unit
+   USE dbcsr_tensor_index, ONLY: &
+      dbcsr_t_get_mapping_info, ndims_mapping, nd_to_2d_mapping
+
+   IMPLICIT NONE
+   PRIVATE
+
+CONTAINS
+
+
+   SUBROUTINE c_f_string(c_str, str)
+      USE, INTRINSIC :: iso_c_binding, ONLY: c_ptr, c_f_pointer, c_char
+      TYPE(c_ptr), INTENT(in) :: c_str
+      CHARACTER(kind=c_char), POINTER :: arr(:)
+      CHARACTER(:, kind=c_char), ALLOCATABLE, INTENT(out) :: str
+      INTEGER(8) :: n, i
+      INTERFACE
+         ! steal std c library function rather than writing our own.
+         FUNCTION strlen(s) bind(c, name='strlen')
+            USE, INTRINSIC :: iso_c_binding, ONLY: c_ptr, c_size_t
+            IMPLICIT NONE
+            !----
+            TYPE(c_ptr), INTENT(in), value :: s
+            INTEGER(c_size_t) :: strlen
+         END FUNCTION strlen
+      END INTERFACE
+      n = strlen(c_str)
+      !****
+      CALL c_f_pointer(c_str, arr, [n])
+      ALLOCATE (CHARACTER(len=n) :: str)
+      DO i = 1, n
+         str(i:i) = arr(i)
+      ENDDO
+   END SUBROUTINE c_f_string
+   
+   SUBROUTINE c_dbcsr_t_finalize(c_tensor) BIND(C, name="c_dbcsr_t_finalize")
+      TYPE(c_ptr), INTENT(IN), VALUE :: c_tensor
+      TYPE(dbcsr_t_type), POINTER :: tensor
+      
+      CALL c_f_pointer(c_tensor, tensor)
+      
+      CALL dbcsr_t_finalize(tensor)
+
+   END SUBROUTINE
+
+   SUBROUTINE c_dbcsr_t_pgrid_create_expert(fcomm, c_dims, dims_size, c_pgrid, &
```

Comment:
I created the `expert` routines for internal use and they should not be exposed to the API. Do you use this? It's part of the Fortran API only for testing with CP2K. I may remove this at a later time.

Reply:
I don't use it, so you can remove it later.